frontends/web/src/routes/buy/utils.tsx

```diff
@@ -0,0 +1,69 @@
+/**
+ * Copyright 2022 Shift Crypto AG
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { ExchangeDeals } from '../../api/exchanges';
+import { ExchangeProviderWithBestDeal, PaymentMethod } from './types';
+
+/**
+ * Finds the lowest fee among all `supported`
+ * exchange providers for a given region.
+ */
+export function findLowestFee(providers: ExchangeProviderWithBestDeal[]) {
+  // all payment methods' fees of all
+  // supported providers (for a selected region)
+  let allFees: number[] = [];
+
+  providers.forEach(provider => {
+    provider.methods.forEach(method => {
+      if (provider.supported) {
+        allFees = [...allFees, method.fee];
+      }
+    });
+  });
+
+  return Math.min(...allFees);
+}
+
+export function restructureExchangeDeals(exchangeDeals: ExchangeDeals): ExchangeProviderWithBestDeal[] {
+  // Loop through each provider in the exchangeDeals object
+  let arr: ExchangeProviderWithBestDeal[] = [];
+  const exchangeDealsKeys = Object.keys(exchangeDeals);
+  exchangeDealsKeys.forEach((key) => {
+    let providerName = key as keyof ExchangeDeals;
+    const providerObj = { providerName, methods: exchangeDeals[providerName] as PaymentMethod[] };
+    arr = [...arr, providerObj];
+  });
+
+  //returns array of object of all providers
+  return arr.map(provider => ({
+    ...provider,
+    methods: provider.methods.sort((a, b) => a.payment.localeCompare(b.payment))
+  }));
+}
+
+export function findBestDeal(providers: ExchangeProviderWithBestDeal[], lowestFee: number): ExchangeProviderWithBestDeal[] {
+  // for each provider's payment method
+  // mark it as "bestDeal" if the
+  // fee is equal to the lowest fee
+  // accross ALL supported providers
+  return providers.map(provider => ({
+    ...provider,
+    methods: provider.methods.map(method => ({
+      ...method,
+      isBestDeal: method.fee === lowestFee && providers.filter(p => p.supported).length > 1
```

Comment:
Filter and length check could be moved at beginning to skip the map if there are less than 2 providers? Just a nit, though